packages/@lwc/engine-dom/src/custom-elements/create-custom-element-scoped.ts

```diff
@@ -0,0 +1,93 @@
+        // Why not lazily patch in `createCustomElement`? Well, this could lead to subtle bugs, e.g.:
+        //
+        // 1. LWC loads
+        // 2. `const Ctor = class extends HTMLElement {}`
+        // 3. `lwc.createElement(...)` // here we lazily patch
+        // 4. `customElements.define('x-foo', Ctor)` // throws error because class is bound to stale HTMLElement
+        //
+        // To reduce the risk of this, it's safer to patch the registry eagerly.
+        createScopedConstructor = createScopedRegistry();
+
+        // It's important to cache window.HTMLElement here. Otherwise, someone else could overwrite window.HTMLElement (e.g.
+        // another copy of the engine, or another scoping implementation) and we would get "Illegal constructor" errors
+        // because the HTMLElement prototypes are mixed up.
+        //
+        // The reason this happens is that the scoping implementation overwrites window.HTMLElement and expects to work
+        // with that version of HTMLElement. So if you load two copies of the scoping implementation in the same environment,
+        // the second one may accidentally grab window.HTMLElement from the first (when doing `class extends HTMLElement`).
+        // Caching avoids this problem.
+        CachedHTMLElement = window.HTMLElement;
+    }
+}
+
+// Creates a constructor that is intended to be used as the UserConstructor in a scoped (pivots) registry.
+// In this case, the upgradeCallback only needs to be defined once because we create these on-demand,
+// multiple times per tag name.
+const createUserConstructor = (
+    upgradeCallback: LifecycleCallback,
+    connectedCallback: LifecycleCallback,
+    disconnectedCallback: LifecycleCallback,
+    HTMLElementToExtend: typeof HTMLElement
+) => {
+    // TODO [#2972]: this class should expose observedAttributes as necessary
+    return class UserConstructor extends HTMLElementToExtend {
+        constructor() {
+            super();
+            upgradeCallback(this);
+        }
+
+        connectedCallback() {
+            connectedCallback(this);
+        }
+
+        disconnectedCallback() {
+            disconnectedCallback(this);
+        }
+    };
+};
+
+export function createCustomElementScoped(
+    tagName: string,
+    upgradeCallback: LifecycleCallback,
+    connectedCallback: LifecycleCallback,
+    disconnectedCallback: LifecycleCallback
+) {
+    if (isUndefined(createScopedConstructor) || isUndefined(CachedHTMLElement)) {
+        // This error should be impossible to hit
+        throw new Error(
+            'The flag ENABLE_SCOPED_CUSTOM_ELEMENT_REGISTRY must be set to true to use this feature'
+        );
+    }
+
+    const UserConstructor = createUserConstructor(
+        upgradeCallback,
+        connectedCallback,
+        disconnectedCallback,
+        CachedHTMLElement
+    );
+    const ScopedConstructor = createScopedConstructor(tagName, UserConstructor);
+    return new ScopedConstructor(UserConstructor);
+}
```

Comment:
I added a bunch of new comments and added the `if` checks at the top. Otherwise the logic is the same as in https://github.com/salesforce/lwc/pull/2724 .